### PR TITLE
Add support for generating SPIR-V assembly from OpenCL C and C++ for OpenCL

### LINKED ISSUE #2743: [REQUEST] Emit SPIR-V assembly for OpenCL C
**Is your feature request related to a problem? Please describe.**
It would be useful to be able to see the SPIR-V assembly when compiling OpenCL C and C++ for OpenCL code.

**Describe the solution you'd like**
OpenCL C and C++ for OpenCL code can be compiled to SPIR-V assembly as follows:
1. Compile to LLVM bitcode using `clang -cc1 -emit-llvm-bc`
2. Translate the LLVM bitcode to a SPIR-V binary using the LLVM/SPIR-V Translator
3. Disassemble the SPIR-V binary using the `spirv-dis` tool, found in the [SPIR-V Tools repo](https://github.com/KhronosGroup/SPIRV-Tools)

**Additional context**
1. This requires #2741 and #2742 to be completed first. I also have changes written for this issue already.
2. We need to add support for the [LLVM/SPIR-V Bi-Directional Translator](https://github.com/KhronosGroup/SPIRV-LLVM-Translator) in the infra repo.
3. We also need to add support for the [SPIR-V Tools repo](https://github.com/KhronosGroup/SPIRV-Tools)

### PR DESCRIPTION
**Pull request by jzzheng22**
Closes #2743.

This PR requires https://github.com/compiler-explorer/infra/pull/590.

### CODE REVIEW

**Review by partouf — CHANGES_REQUESTED** [review]

File: lib/compilers/spirv.js [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+import { logger } from '../logger';
+import * as utils from '../utils';
+
+export class SPIRVCompiler extends BaseCompiler {
+    static get key() { return 'spirv'; }
+
+    constructor(compilerInfo, env) {
+        super(compilerInfo, env);
+        this.translatorPath = '/opt/compiler-explorer/llvm-spirv-trunk/bin/llvm-spirv';
+        this.disassemblerPath = '/opt/compiler-explorer/SPIRV-Tools-master/build/tools/spirv-dis';
```
[the remark this answers is not in the document]
Comment (jzzheng22) [reply lib/compilers/spirv.js]: Thanks!

**Review by partouf — APPROVED** [review]

**Comment by mattgodbolt** [comment]
Oh my! I realise I hadn't pressed "send" on my review, but like a mind-reader @partouf you have done them all :)

**Comment by mattgodbolt** [comment]
Happy to merge this! Sorry it took so long!

**Comment by mattgodbolt** [comment]
This is now live, thank you!

**Pull request merged**